RecoHGCal/TICL/plugins/TrackstersProducer.cc

```diff
@@ -0,0 +1,103 @@
+#include <vector>
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/CaloRecHit/interface/CaloCluster.h"
+#include "DataFormats/ParticleFlowReco/interface/PFCluster.h"
+
+#include "RecoHGCal/TICL/interface/PatternRecognitionAlgoBase.h"
+#include "RecoHGCal/TICL/interface/Trackster.h"
+#include "PatternRecognitionbyCA.h"
+#include "PatternRecognitionbyMultiClusters.h"
+
+class TrackstersProducer : public edm::stream::EDProducer<> {
+ public:
+  TrackstersProducer(const edm::ParameterSet &);
+  ~TrackstersProducer() override {}
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+
+  void produce(edm::Event &, const edm::EventSetup &) override;
+
+ private:
+  edm::EDGetTokenT<std::vector<reco::CaloCluster>> clusters_token_;
+  edm::EDGetTokenT<std::vector<std::pair<unsigned int, float>>> filtered_layerclusters_mask_token_;
+  edm::EDGetTokenT<std::vector<float>> original_layerclusters_mask_token_;
+
+  std::unique_ptr<PatternRecognitionAlgoBase> myAlgo_;
+};
+DEFINE_FWK_MODULE(TrackstersProducer);
+
+TrackstersProducer::TrackstersProducer(const edm::ParameterSet& ps)
+    : myAlgo_(std::make_unique<PatternRecognitionbyCA>(ps)) {
+  clusters_token_ = consumes<std::vector<reco::CaloCluster>>(
+      ps.getParameter<edm::InputTag>("hgcal_layerclusters"));
+  filtered_layerclusters_mask_token_ = consumes<std::vector<std::pair<unsigned int, float>>>(
+      ps.getParameter<edm::InputTag>("filtered_layerclusters_mask"));
+  original_layerclusters_mask_token_ =
+      consumes<std::vector<float>>(ps.getParameter<edm::InputTag>("original_layerclusters_mask"));
+  produces<std::vector<Trackster>>("TrackstersByCA");
+  produces<std::vector<float>>();  // Mask to be applied at the next iteration
+}
+
+void TrackstersProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // hgcalMultiClusters
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("hgcal_layerclusters", edm::InputTag("hgcalLayerClusters"));
+  desc.add<edm::InputTag>("filtered_layerclusters_mask",
+                          edm::InputTag("FilteredLayerClusters", "iterationLabelGoesHere"));
+  desc.add<edm::InputTag>("original_layerclusters_mask",
+                          edm::InputTag("hgcalLayerClusters", "InitialLayerClustersMask"));
+  desc.add<int>("algo_verbosity", 0);
+  desc.add<double>("min_cos_theta", 0.915);
+  desc.add<double>("min_cos_pointing", -1.);
+  desc.add<int>("missing_layers", 0);
+  desc.add<int>("min_clusters_per_ntuplet", 10);
+  descriptions.add("tracksters", desc);
```

Comment:
if this ```TrackstersProducer``` module is going to be used in multiple iterations, apparently there is not going to be a unique ```tracksters``` useful for all users.
So, it's more appropriate to stay with the default module name ```trackstersProducer```  for the auto-generated cfi and then clone to an appropriate iteration-driven name in the actual workflow